python/cuml/test/test_kernel_density.py

```diff
@@ -0,0 +1,150 @@
+#
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from cuml.neighbors import KernelDensity, VALID_KERNELS, logsumexp_kernel
+from cuml.common.exceptions import NotFittedError
+from sklearn.metrics import pairwise_distances as skl_pairwise_distances
+from sklearn.neighbors._ball_tree import kernel_norm
+import numpy as np
+from hypothesis import given, settings, assume, strategies as st
+from hypothesis.extra.numpy import arrays
+import pytest
+from sklearn.model_selection import GridSearchCV
+
+
+# not in log probability space
+def compute_kernel_naive(Y, X, kernel, metric, h, sample_weight):
+    d = skl_pairwise_distances(Y, X, metric)
+    norm = kernel_norm(h, X.shape[1], kernel)
+
+    if kernel == "gaussian":
+        k = np.exp(-0.5 * (d * d) / (h * h))
+    elif kernel == "tophat":
+        k = (d < h)
+    elif kernel == "epanechnikov":
+        k = ((1.0 - (d * d) / (h * h)) * (d < h))
+    elif kernel == "exponential":
+        k = (np.exp(-d / h))
+    elif kernel == "linear":
+        k = ((1 - d / h) * (d < h))
+    elif kernel == "cosine":
+        k = (np.cos(0.5 * np.pi * d / h) * (d < h))
+    else:
+        raise ValueError("kernel not recognized")
+    return norm*np.average(k, -1, sample_weight)
+
+
+@st.composite
+def array_strategy(draw):
+    n = draw(st.integers(1, 100))
+    m = draw(st.integers(1, 100))
+    dtype = draw(st.sampled_from([np.float64, np.float32]))
+    rng = np.random.RandomState(34)
+    X = rng.randn(n, m).astype(dtype)
+    n_test = draw(st.integers(1, 100))
+    X_test = rng.randn(n_test, m).astype(dtype)
+
+    if draw(st.booleans()):
+        sample_weight = None
+    else:
+        sample_weight = draw(arrays(dtype=np.float64, shape=n,
+                                    elements=st.floats(0.1, 2.0),))
+
+    return X, X_test, sample_weight
+
+
+metrics_strategy = st.sampled_from(
+    ['euclidean', 'manhattan',
+     'chebyshev', 'minkowski', 'hamming', 'canberra'])
+
+
+@settings(deadline=None, max_examples=100)
+@given(array_strategy(), st.sampled_from(VALID_KERNELS),
```

Comment:
Can we test this w/ numpy, cupy, cudf, and pandas df as input? I might have missed this in the kernel ridge PR. We probably want to do this there as well just to remove potential surprises.

Reply:
I started doing this but it has opened up more potential bugs that I need to work through. Cuml's automatic type conversion is sometimes converting 2 dimensional cudf dataframes to one dimension.

Reply:
Ok I now have both KDE and KRR tested with all inputs.